### ISSUE TITLE
STT enters infinite repetition loop when input audio contains repeated words - missing audio repetition penalty

### ISSUE THREAD

**Issue by damienlaine** [kyutai-labs/delayed-streams-modeling#175]
## Description

When the input audio contains repeated words (e.g., someone stuttering or intentionally repeating a word), the STT model gets stuck in an infinite loop, continuously outputting the same word indefinitely instead of continuing with the rest of the transcription.

## Environment

- **Server:** `moshi-server worker` (Rust)
- **Model:** `kyutai/stt-1b-en_fr-candle`
- **Config:** Standard `config-stt-en_fr-hf.toml`
- **GPU:** NVIDIA (CUDA)
- **Docker base:** `nvidia/cuda:12.9.1-runtime-ubuntu22.04`

## Configuration

```toml
[modules.asr]
path = "/api/asr-streaming"
type = "BatchedAsr"
lm_model_file = "hf://kyutai/stt-1b-en_fr-candle/model.safetensors"
text_tokenizer_file = "hf://kyutai/stt-1b-en_fr-candle/tokenizer_en_fr_audio_8000.model"
audio_tokenizer_file = "hf://kyutai/stt-1b-en_fr-candle/mimi-pytorch-e351c8d8@125.safetensors"
asr_delay_in_tokens = 6
batch_size = 64
temperature = 0.0

[modules.asr.model.transformer]
context = 750
max_seq_len = 40960
```

## Steps to Reproduce

1. Start `moshi-server worker` with the STT config
2. Stream audio containing repeated words (e.g., "hello hello hello hello hello")
3. Observe the transcription output

## Expected Behavior

The model should transcribe the repeated words and then continue transcribing the rest of the audio normally.

## Actual Behavior

The model enters an infinite loop, outputting the repeated word continuously (e.g., "hello hello hello hello hello hello hello...") and never recovers, even after the audio input has moved on to different content.

## Root Cause Analysis

The `repetition_penalty` mechanism exists for **text tokens** but is **completely missing for audio tokens** in both Python and Rust implementations.

### Python (PyTorch) - `moshi/moshi/utils/sampling.py`

The `sample_token` function has no `repetition_penalty` parameter:

```python
def sample_token(
    logits: torch.Tensor,
    use_sampling: bool = False,
    temp: float = 1.0,
    top_k: int = 0,
    top_p: float = 0.0,
    # ❌ MISSING: past_tokens, repetition_penalty
) -> torch.Tensor:
    if use_sampling and temp > 0.0:
        probs = torch.softmax(logits / temp, dim=-1)
        # ... no penalty applied
```

### Rust - `moshi-core/src/lm.rs` - `DepFormer::sample()`

Audio logits go directly to `LogitsProcessor.sample()` without any penalty:

```rust
// For each codebook slice:
let logits = xs.apply(&slice.linear_out)?;
let token = lp.sample(&logits)?;  // ❌ No repetition penalty before sampling
```

### Rust - `moshi-core/src/lm_generate_multistream.rs`

The `apply_repetition_penalty()` method exists but is **only called for text tokens**, not audio:

```rust
fn apply_repetition_penalty(&self, logits: Tensor) -> candle::Result<Tensor> {
    // Only iterates over self.text_tokens
    for &token_id in self.text_tokens(false).iter().rev() {
        // ... penalty logic
    }
}
```

Audio tokens are stored in `audio_tokens: Vec<Vec<u32>>` but never used for penalty calculation.

### Python - Add penalty to `sample_token()`:

```python
def sample_token(
    logits: torch.Tensor,
    use_sampling: bool = False,
    temp: float = 1.0,
    top_k: int = 0,
    top_p: float = 0.0,
    past_tokens: torch.Tensor = None,      # NEW
    repetition_penalty: float = 1.0,        # NEW
) -> torch.Tensor:
    # Apply repetition penalty BEFORE temperature scaling
    if past_tokens is not None and repetition_penalty != 1.0:
        for token in past_tokens:
            if logits[..., token] > 0:
                logits[..., token] /= repetition_penalty
            else:
                logits[..., token] *= repetition_penalty
    
    if use_sampling and temp > 0.0:
        probs = torch.softmax(logits / temp, dim=-1)
        # ...
```

### Rust - Add penalty to `DepFormer::sample()`:

```rust
fn sample(
    &mut self,
    ys: &Tensor,
    text_token: Option<u32>,
    forced: &[Option<u32>],
    lp: &mut LogitsProcessor,
    past_audio_tokens: Option<&[Vec<u32>]>,  // NEW
    audio_penalty: Option<f32>,               // NEW
) -> Result<Vec<u32>> {
    for (codebook_idx, slice) in self.slices.iter_mut().enumerate() {
        let mut logits = xs.apply(&slice.linear_out)?;
        
        // NEW: Apply audio repetition penalty per codebook
        if let (Some(past), Some(penalty)) = (&past_audio_tokens, audio_penalty) {
            if let Some(codebook_history) = past.get(codebook_idx) {
                logits = apply_audio_penalty(logits, codebook_history, penalty)?;
            }
        }
        
        let token = lp.sample(&logits)?;
        // ...
    }
}
```

## Additional Context

- Setting `temperature = 0.0` (deterministic decoding) does not prevent this issue
- This is a **blocking issue** for production use cases where users may stutter or repeat words naturally

## Related Issues

- #146 - STT sometimes stops transcribing (possibly related)
